### PR TITLE
Prototype of Scraper

### PR DESCRIPTION
**Pull request by kelwang**
This PR will add a scrapper service in influx.
It will accept one prometheus URL together with orgID and bucketID, send to Nats queue for parsing. After the parsing, send to Nats for storing.

This PR also have gather lib, which includes interfaces and some implementations, and unit test cases. The influxStorage Record method will be implemented later, currently will just print out for testing. 

Usage Example to be updated here

### CODE REVIEW

File: gather/metrics.go [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+
+import dto "github.com/prometheus/client_model/go"
+
+// Metrics is the default influx based metrics
+type Metrics struct {
+	Name      string                 `json:"name"`
+	Tags      map[string]string      `json:"tags"`
+	Fields    map[string]interface{} `json:"fields"`
+	Timestamp int64                  `json:"timestamp"`
+	Type      dto.MetricType         `json:"type"`
```
Comment (goller) [inline gather/metrics.go]: Can you write this so that the prometheus import does not leak out?

File: gather/scrapper.go [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+	"time"
+
+	"github.com/influxdata/platform"
+	"github.com/influxdata/platform/nats"
+	"github.com/matttproud/golang_protobuf_extensions/pbutil"
+	dto "github.com/prometheus/client_model/go"
+	"github.com/prometheus/common/expfmt"
+)
+
+/*
```
Comment (goller) [inline gather/scrapper.go]: remove this comment

File: gather/scrapper.go [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+func (p *prometheusScrapper) parse(r io.Reader, header http.Header) error {
+	var parser expfmt.TextParser
+
+	mediatype, params, err := mime.ParseMediaType(header.Get("Content-Type"))
+	if err != nil {
+		return err
+	}
+	// Prepare output
+	metricFamilies := make(map[string]*dto.MetricFamily)
+	if err == nil && mediatype == "application/vnd.google.protobuf" &&
```
Comment (goller) [inline gather/scrapper.go]: will err ever be no nil here?

File: cmd/scrapperd/main.go [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+		HostURL:  pURL,
+		OrgID:    *orgID,
+		BucketID: *bucketID,
+	}
+	return msg, influxURL
+}
+
+func initNats(logger *log.Logger, influxURL []string) nats.Publisher {
+	server := nats.NewServer(nats.Config{
+		FilestoreDir: ".",
```
Comment (jademcgough) [inline cmd/scrapperd/main.go]: This shouldn't use the current directory for the file store.

Comment (kelwang) [reply cmd/scrapperd/main.go]: 👍

File: gather/scrapper.go [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+/*
+	1. Discoverer
+	2. Scheduler
+	3. Queue
+	4. Scrap
+	5. Storage
+*/
+
+// Scrapper gathers metrics from an url
+type Scrapper interface {
```
[the remark this answers is not in the document]
Comment (kelwang) [reply gather/scrapper.go]: 👍

File: cmd/scrapperd/main.go [hunk 6]
```diff
@@ -0,0 +1,114 @@
+package main
```
[the remark this answers is not in the document]
Comment (kelwang) [reply cmd/scrapperd/main.go]: 👍

File: gather/scrapper.go (hunk 2 above)
Comment (kelwang) [reply gather/scrapper.go]: 👍

File: http/scraper_service.go [hunk 7]
```diff
@@ -0,0 +1,214 @@
...
+func (h *ScraperHandler) handleGetScraperTargets(w http.ResponseWriter, r *http.Request) {
+	ctx := r.Context()
+
+	targets, err := h.ScraperStorageService.ListTargets(ctx)
+	if err != nil {
+		EncodeError(ctx, err, w)
+		return
+	}
+
+	if err := encodeResponse(ctx, w, http.StatusOK, targets); err != nil {
```
Comment (goller) [inline http/scraper_service.go]: This encoding should be

```json
{
    "links": {
          "next": "/api/v2/metrics????&page=2",
    },
    "targets": [
    {
        "name": "myname",
        "id": "1234",
        "org": "myorg",
        "bucket": "mybucket",
        "type": "prometheus",
        "url": "http://127.0.0.1:9090/metrics",
       "links": {
            "self": "/api/v2/?metrics?/1234"
      }

    }
]
}
```

File: service.go [hunk 8]
```diff
@@ -0,0 +1,9 @@
+package platform
```
Comment (goller) [inline service.go]: Sorry... I told you to do this... I was wrong.

File: metrics.go [hunk 9]
```diff
@@ -0,0 +1,56 @@
+package platform
```
Comment (goller) [inline metrics.go]: this is an implementation detail of the prometheus scraper.  Would you move this into scraper?

File: scraper.go [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+
+// ScraperTargetFilter represents a set of filter that restrict the returned results.
+type ScraperTargetFilter struct {
+	ID   *ID     `json:"id"`
+	Name *string `json:"name"`
+}
+
+// ScraperTargetUpdate represents updates to a scraper target.
+// Only fields which are set are updated.
+type ScraperTargetUpdate struct {
```
Comment (goller) [inline scraper.go]: This is a detail of http only.  This struct should be moved into the HTTP handler

File: scraper.go [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+	// PrometheusScraperType parses metrics from a prometheus endpoint.
+	PrometheusScraperType = "prometheus"
+)
+
+var validScraperTypes = map[ScraperType]bool{
+	PrometheusScraperType: false,
+}
+
+// ValidScraperType returns true is the type string is valid
+func ValidScraperType(s string) bool {
```
Comment (goller) [inline scraper.go]: No need for the map right now.  Just a simple switch.

File: scraper.go [hunk 12]
```diff
@@ -0,0 +1,62 @@
+package platform
+
+import (
+	"context"
+)
+
+// Scraper gathers metrics from an url
+type Scraper interface {
+	Gather(ctx context.Context, orgID, BucketID ID, url string) (ms []Metrics, err error)
```
Comment (goller) [inline scraper.go]: Gather should just take a `ScraperTarget`

**Pull request merged**